common/authorization/rsaTokenKeyProvider.go

```diff
@@ -0,0 +1,150 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorization
+
+import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/loggerimpl"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/service/config"
+)
+
+type jwks struct {
+	Keys []jsonWebKeys `json:"keys"`
+}
+
+type jsonWebKeys struct {
+	Kty string   `json:"kty"`
+	Kid string   `json:"kid"`
+	Use string   `json:"use"`
+	N   string   `json:"n"`
+	E   string   `json:"e"`
+	X5c []string `json:"x5c"`
+}
+
+// Default RSA token key provider
+type rsaKeyProvider struct {
+	config config.JWTKeyProvider
+	keys   map[string]*rsa.PublicKey
+	timer  *time.Timer
+	logger log.Logger
+}
+
+var _ TokenKeyProvider = (*rsaKeyProvider)(nil)
+
+func NewRSAKeyProvider(cfg *config.Config) *rsaKeyProvider {
+	logger := loggerimpl.NewLogger(cfg.Log.NewZapLogger())
+	provider := rsaKeyProvider{config: cfg.Global.Security.JWTKeyProvider, logger: logger}
+	provider.init()
+	return &provider
+}
+
+func (a *rsaKeyProvider) init() {
+	a.keys = make(map[string]*rsa.PublicKey)
+	if len(a.config.KeySourceURIs) == 0 {
+		return // nothing to do
+	}
+	err := a.updateKeys()
+	if err != nil {
+		a.logger.Error("error during initial retrieval of RSA token keys: ", tag.Error(err))
+	}
+	a.timer = time.AfterFunc(a.config.RefreshTime, func() {
+		a.timerCallback()
+	})
+}
+
+func (a *rsaKeyProvider) rsaKey(alg string, kid string) (*rsa.PublicKey, error) {
+	if strings.ToLower(alg) != "rs256" {
+		return nil, fmt.Errorf("unexpected signing algorithm: %s", alg)
+	}
+
+	key, found := a.keys[kid]
+	if !found {
+		return nil, fmt.Errorf("RSA key not found for key ID: %s", kid)
+	}
+	return key, nil
+}
+
+func (a *rsaKeyProvider) timerCallback() {
+	err := a.updateKeys()
+	if err != nil {
+		a.logger.Error("error while refreshing RSA token keys: ", tag.Error(err))
+	}
+	a.timer = time.AfterFunc(a.config.RefreshTime, func() {
+		a.timerCallback()
+	})
+}
+
+func (a *rsaKeyProvider) updateKeys() error {
+	if len(a.config.KeySourceURIs) == 0 {
+		return fmt.Errorf("no URIs configured for retrieving keys")
+	}
+
+	keys := make(map[string]*rsa.PublicKey)
+
+	for _, uri := range a.config.KeySourceURIs {
+		resp, err := http.Get(uri)
```

Comment:
interestingly, if one of the URIs is bad, but the rest are good, we won't do the update of the key dictionary. Is that the intended behavior we want?

Reply:
The problem I see here is that an update both potentially brings new keys *and* removes old ones. If the data we got is incomplete because of a failure to call one of the URIs and we replace the map, we'll lose the existing keys for the failed URI that are most likely still valid, and we won't be able to validate tokens until we get a successful refresh from the failed URI. So, for transient failures, I think it's better to do all or nothing.

However, if the error isn't transient, then this approach will block refreshing of all keys from all sources, which would be bad. This makes me think that keys should be organized by source URIs, not as a flat map.

Are you okay if I leave it as is for now and do a follow-up PR for that?

Reply:
sounds good to me

Reply:
yeah, sounds good to me